src/common/components/Button/Button.tsx

```diff
@@ -0,0 +1,146 @@
+import Link from "next/link";
+import {
+  type ComponentPropsWithoutRef,
+  type ReactNode,
+  useCallback,
+  isValidElement,
+  Children,
+  type HTMLProps,
+  cloneElement,
+  type ReactElement,
+} from "react";
+
+import {
+  buttonIconTheme,
+  buttonTheme,
+  type ButtonVariants,
+} from "./Button.theme";
+
+interface ButtonBaseProps {
+  iconLeft?: ReactNode;
+  iconRight?: ReactNode;
+  isResponsive?: boolean;
+}
+
+export interface ButtonLinkProps extends ComponentPropsWithoutRef<"a"> {
+  as?: "a";
+  external?: boolean;
+  href: string;
+}
+
+export interface ButtonProps extends ComponentPropsWithoutRef<"button"> {
+  as?: "button";
+}
+
+// Discriminated union based on "as" prop
+export type ButtonOrLinkProps = (ButtonLinkProps | ButtonProps) &
+  ButtonBaseProps &
+  Omit<ButtonVariants, "hasIcon" | "iconOnly">;
+
+export const Button = ({
+  children,
+  iconLeft,
+  iconRight,
+  isResponsive = false,
+  ...props
+}: ButtonOrLinkProps) => {
+  // Conditionally render between <NextLink>, <a> or <button> depending on props
+  // useCallback to prevent unnecessary re-rendering
+  const Component = useCallback(
+    ({ children: _children, ..._props }: ButtonOrLinkProps) => {
+      // Accessing .as instead of destructuring to make use of discriminated unions
+      if (_props.as === "a") {
```

Comment:
bummer that type narrowing on non-common discriminated union properties will result in type errors...

this issue is currently tracked at `https://github.com/microsoft/TypeScript/issues/46318` and tagged as awaiting more feedback, so it might be worth to include this link in the code as a reminder to update this part of the code when it eventually gets included in future typescript versions!